compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaLexer.java

```diff
@@ -1550,4 +1561,271 @@ private STToken readTokenInBracedContentInInterpolation() {
+     */
+
+    private STToken readDocumentationToken() {
+        reader.mark();
+        int nextChar = peek();
+        switch (nextChar) {
+            case LexerTerminals.PLUS:
+                reader.advance();
+                switchMode(ParserMode.DOCUMENTATION_PARAMETER);
+                return getDocumentationSyntaxToken(SyntaxKind.PLUS_TOKEN);
+            case LexerTerminals.BACKTICK:
+                if (reader.peek(1) != LexerTerminals.BACKTICK) {
+                    reader.advance();
+                    switchMode(ParserMode.DOCUMENTATION_BACKTICK_CONTENT);
+                    return getSyntaxToken(SyntaxKind.BACKTICK_TOKEN);
+                }
+                // Fall through
+            default:
+                while (!reader.isEOF()) {
+                    switch (nextChar) {
+                        case LexerTerminals.NEWLINE:
+                        case LexerTerminals.CARRIAGE_RETURN:
+                            endMode();
+                            break;
+                        case LexerTerminals.BACKTICK:
+                            if (reader.peek(1) != LexerTerminals.BACKTICK) {
+                                break;
+                            } else if (reader.peek(2) != LexerTerminals.BACKTICK) {
+                                // Double backtick detected
+                                reader.advance(2);
+                                advanceReaderForBacktickContent(false);
+                            } else {
+                                // Triple backtick detected
+                                reader.advance(3);
+                                advanceReaderForBacktickContent(true);
+                            }
+                            nextChar = peek();
+                            continue;
+                        default:
+                            if (isIdentifierInitialChar(nextChar)) {
+                                // Look ahead and see if next characters belong to a documentation reference.
+                                // If they do, switch the mode immediately and return.
+                                // Otherwise advance the reader for checked characters.
+                                int readerAdvanceCount = lookAheadForDocumentationReference(nextChar);
+                                if (readerAdvanceCount == 0) {
+                                    switchMode(ParserMode.DOCUMENTATION_REFERENCE_TYPE);
+                                    if (getLexeme().equals("")) {
+                                        // Reaching here means,
+                                        // first immediate character also belong to a documentation reference.
+                                        return readDocumentationReferenceTypeToken();
+                                    }
+                                    break;
+                                }
+                                reader.advance(readerAdvanceCount);
```

Comment:
Is it possible to merge this section with the `lookAheadForDocumentationReference` method?

i.e:  lookAheadForDocumentationReference will process all the tokens and returns the token, instead of the lookahead count